File: app/src/main/java/com/example/jetpackmvvm/sample/ui/repo/RepoViewModel.java
package com.example.jetpackmvvm.sample.ui.repo;

import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.example.jetpackmvvm.sample.data.model.Repo;
import com.example.jetpackmvvm.sample.util.Event;

import java.util.List;

public class RepoViewModel extends ViewModel {

    public final MutableLiveData<Event<List<Repo>>> repos = new MutableLiveData<>();
    public final MutableLiveData<Boolean> isLoading = new MutableLiveData<>();
    private RepoDataModel repoDataModel;

    public RepoViewModel() {
        init(new RepoDataModel());
    }

    public RepoViewModel(RepoDataModel dataModel) {
        init(dataModel);
    }

    private void init(RepoDataModel dataModel) {
        repoDataModel = dataModel;
    }

    public void searchRepo(String query) {
        isLoading.setValue(true);
        repoDataModel.searchRepo(query, new RepoDataModel.onDataReadyCallback() {
            @Override
            public void onDataReady(List<Repo> data) {
                RepoViewModel.this.repos.setValue(new Event<>(data));
                isLoading.setValue(false);
            }
        });
    }
}

File: app/src/main/java/com/terry/jetpackmvvm/sample/ui/repo/RepoViewModel.kt
package com.terry.jetpackmvvm.sample.ui.repo

import android.text.TextUtils
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Transformations
import androidx.lifecycle.ViewModel
import com.terry.jetpackmvvm.sample.data.model.Repo
import com.terry.jetpackmvvm.sample.util.AbsentLiveData
import com.terry.jetpackmvvm.sample.util.Event
import javax.inject.Inject

class RepoViewModel : ViewModel {
    @Inject
    lateinit var repoRepository: RepoRepository
    private val _isLoading = MutableLiveData<Boolean>()
    private val query = MutableLiveData<String>()
    private var _repos: LiveData<Event<List<Repo>>> = Transformations.switchMap(query) { input ->
        if (TextUtils.isEmpty(input)) {
            AbsentLiveData.create()
        } else {
            repoRepository.searchRepo(input)
        }
    }

    val repos: LiveData<Event<List<Repo>>> = _repos
    val isLoading: LiveData<Boolean> = _isLoading

    @Inject
    constructor() {
//        init(RepoRepository())
    }

    constructor(repoRepository: RepoRepository) {
        init(repoRepository)
    }

    private fun init(repoRepository: RepoRepository) {
        this.repoRepository = repoRepository
    }

    fun searchRepo(inputText: String) {
        query.value = inputText
    }

    fun setLoading(value: Boolean) {
        _isLoading.value = value
    }

    override fun onCleared() {
        super.onCleared()
    }
}